mpl_interactions/xarray_helpers.py

```diff
@@ -0,0 +1,159 @@
+import numpy as np
+
+
+def choose_fmt_str(dtype=None):
+    """
+    Choose the appropriate string formatting for different dtypes.
+
+    Paramters
+    ---------
+
+    dtype: np.dtye
+        Dtype of array containing values to be formatted.
+
+
+    Returns
+    -------
+
+    fmt: str
+        Format string
+
+    """
+    if np.issubdtype(dtype, "float"):
+        fmt = r"{:0.2f}"
+    elif np.issubdtype(dtype, "int"):
+        fmt = r"{:d}"
+    else:
+        fmt = r"{:}"
+    return fmt
+
+
+def choose_datetime_nonsense(arr, timeunit="m"):
+    """
+    Try to do something reasonable to datetimes and timedeltas.
+
+    Parameters
+    ----------
+
+    arr: np.array
+        Array with values to be formatted.
+
+    Returns
+    -------
+
+    out: np.array
+        Array modified to format decently in a slider.
+
+    """
+    if np.issubdtype(arr.dtype, "datetime64"):
+        # print('datetime')
+        out = arr.astype(f"datetime64[{timeunit}]")
+    elif np.issubdtype(arr.dtype, "timedelta64"):
+        out = arr.astype(f"timedelta64[{timeunit}]").astype(int)
+    else:
+        out = arr
+    return out
+
+
+def get_hs_axes(xarr, is_color_image=False, timeunit="m"):
+    """
+    Read the dims and coordinates from an xarray and construct the
+    axes argument for hyperslicer. Called internally by hyperslicer.
+
+    Parameters
+    ----------
+
+    xarr: xarray.DataArray
+        DataArray being viewed with hyperslicer
+
+    is_color_image: bool default False
+        Whether the individual images of the hyperstack are color images.
+
+    timeunit: str default "m"
+        Truncation level for datetime and timedelta axes.
+
+    Returns
+    -------
+    axes: list
+        axes kwarg for hyperslicer
+
+    """
+    if not is_color_image:
+        dims = xarr.dims[:-2]
+    else:
+        dims = xarr.dims[:-3]
+    coords_list = [choose_datetime_nonsense(xarr.coords[d].values, timeunit=timeunit) for d in dims]
+    # print(coords_list)
+    axes = zip(dims, coords_list)
+    return list(axes)
+
+
+def get_hs_extent(xarr, is_color_image=False):
+    """
+    Read the "XY" coordinates of an xarray.DataArray to set extent of image for
+    imshow.
+
+    Parameters
+    ----------
+
+    xarr: xarray.DataArray
+        DataArray being viewed with hyperslicer
+
+    is_color_image: bool default False
+        Whether the individual images of the hyperstack are color images.
+
+    Returns
+    -------
+    extent: list
+        Extent argument for imshow. [d0_min, d0_max, d1_min, d1_max]
+
+    """
+
+    if not is_color_image:
+        dims = xarr.dims[-2:]
+    else:
+        dims = xarr.dims[-3:-1]
+    extent = []
+    for d in dims:
```

Comment:
will this break if I give an xarray that doesn't have any dims?

Reply:
no it doesn't. but i don't really understand why it doesn't. It seems like it should have to return `None` in that case. but i guess maptlotlib handles `[]` correctly?

Reply:
xarrays have to have have dims? Or maybe its handled by the part where we check that the input is at least 3d? not sure i get how you would end up with dim-less xarray.

Reply:
i wasn't sure what would happen if they didn't have explcitiyl named dims, but seems to work, so all is well.